sycl/source/detail/kernel_bundle_impl.hpp

```diff
@@ -396,6 +397,44 @@ class kernel_bundle_impl {
     return SS.str();
   }
 
+  bool
+  extKernelCompilerFetchFromCache(const std::vector<device> Devices,
+                                  const std::vector<std::string> &BuildOptions,
+                                  const std::string &SourceStr,
+                                  ur_program_handle_t &UrProgram) {
+    using ContextImplPtr = std::shared_ptr<sycl::detail::context_impl>;
+    ContextImplPtr ContextImpl = getSyclObjImpl(MContext);
+    const AdapterPtr &Adapter = ContextImpl->getAdapter();
+
+    std::string UserArgs = syclex::detail::userArgsAsString(BuildOptions);
+    auto BinProg = PersistentDeviceCodeCache::getCompiledKernelFromDisc(
+        Devices[0], UserArgs, SourceStr);
+    if (!BinProg.empty()) {
+      ur_device_handle_t UrDevice = getSyclObjImpl(Devices[0])->getHandleRef();
+      ur_result_t BinaryStatus = UR_RESULT_SUCCESS;
+      ur_program_properties_t Properties = {};
+      std::vector<ur_program_metadata_t> Metadata = {};
+      Properties.stype = UR_STRUCTURE_TYPE_PROGRAM_PROPERTIES;
+      Properties.pNext = nullptr;
+      Properties.count = Metadata.size();
+      Properties.pMetadatas = Metadata.data();
+      BinaryStatus = Adapter->call_nocheck<UrApiKind::urProgramCreateWithBinary>(
+          ContextImpl->getHandleRef(), UrDevice, BinProg[0].size(),
+          (const unsigned char *)BinProg[0].data(), &Properties, &UrProgram);
+
+      if (BinaryStatus == UR_RESULT_SUCCESS) {
+        ur_result_t Error = Adapter->call_nocheck<UrApiKind::urProgramBuildExp>(
+            UrProgram,
+            /*num devices =*/1, &UrDevice, UserArgs.c_str());
```

Comment:
Why do we only use one device here? I see that `urProgramCreateWithBinary` only takes a single device, but should we then do this for each device?

Reply:
This is consistent with how the persistent cache works now. But let me look into it. 
`urProgramBuildExp` can take more devices, we aren't limited to one.

Reply:
@steffenlarsen - Artur has https://github.com/oneapi-src/unified-runtime/pull/2147  out there to make `urProgramCreateWithBinary` work with multiple devices, and a follow on to pull that into the UR. 
I think its best is to simply wait on that work to get merged and then update this. 
If there is a delay or some other deadline issue, then we can merge my PR as is and follow up later for multiple devices, but today I think we should just wait.